torch/csrc/jit/backends/backend_init.cpp

```diff
@@ -2,11 +2,83 @@
 #include <torch/csrc/jit/backends/backend_detail.h>
 #include <torch/csrc/jit/backends/backend_resolver.h>
 #include <torch/csrc/jit/frontend/code_template.h>
+#include <torch/csrc/jit/python/module_python.h>
 #include <torch/csrc/jit/python/pybind_utils.h>
 
 namespace torch {
 namespace jit {
 
+// Selectively lower \p cloned_mod to a backend. \p to_backend
+// is called to lower modules. \p modules_to_lower contains
+// qualified names of submodules of \p cloned_mod that should be lowered.
+void to_backend_selective_impl(
+    Module& cloned_mod,
+    py::function to_backend,
+    const std::vector<std::string>& modules_to_lower) {
+  // For each module that should be lowered:
+  for (const auto& module_to_lower : modules_to_lower) {
+    // Use QualifiedName to parse the qualified module names.
+    c10::QualifiedName qual_module_name(module_to_lower);
+    auto& atoms = qual_module_name.atoms();
+
+    // Search through the module hierarchy using the atoms of
+    // qual_module_name until current points to the module to
+    // be lowered. Make parent point to its parent for type
+    // modification purposes.
+    Module parent;
+    Module current = cloned_mod;
+    for (size_t i = 0, e = atoms.size(); i < e; ++i) {
+      IValue submodule = current.attr(atoms[i]);
+      if (submodule.isModule()) {
+        // If this is the last atom, store a reference to the
+        // parent of the module that is about to be lowered.
+        if (i == e - 1) {
+          parent = current;
+        }
+        current = submodule.toModule();
+      } else {
+        std::stringstream err;
+        err << "Attribute named " << atoms[i] << " is not a Module";
+        throw std::runtime_error(err.str());
+      }
+    }
+
+    // Call to_backend on the module that needs to be lowered. It needs to be
+    // wrapped before doing so because _to_jit_backend accepts wrapped modules.
+    // The result needs to be unwrapped in order to access its type below.
+    auto lowered_submodule =
+        py::cast<Module>(to_backend(py::module::import("torch.jit._recursive")
+                                        .attr("wrap_cpp_module")(current))
+                             .attr("_c"));
+
+    // Adjust the parent's type so that the type of the submodule matches
+    // the type of lowered_submodule.
+    auto parent_type = parent.type();
+    parent_type->unsafeChangeAttributeType(
+        qual_module_name.name(), lowered_submodule.type());
+    parent.setattr(qual_module_name.name(), lowered_submodule._ivalue());
+
+    // Fix references to the submodule in the graphs of parent's methods.
+    std::unordered_map<TypePtr, TypePtr> type_remap;
+    type_remap[current.type()] = lowered_submodule.type();
+
+    auto type_remap_fn = [&type_remap](TypePtr in) {
+      auto it = type_remap.find(in);
+      if (it == type_remap.end())
+        return in;
+      return it->second;
+    };
+
+    for (auto& fn : parent_type->methods()) {
```

Comment:
Still not totally sure about the type remapping logic here. For example, this assumes that only the direct parent of a module needs to have the graphs fixed up, but a module can appear in the graph of any of its recursive parents (through an attribute access like: `self.nested1.nested2.lowered_module`).

So I think to get all the corner cases, it's best to specify what should happen top-down and apply the transformation globally. e.g. build up the typemap of all modules to be lowered, the walk the module hierarchy and rewrite all methods/slots according to that map.

Reply:
I modified `to_backend_selective_impl` so that it maintains references to all of the ancestors of the lowered module, fixes up the parent type (the type of the direct ancestor), and then remaps types the graphs and schemas of all of the methods of all of its ancestors.

Do we need to worry about rewriting slots in any module except the parent? A module can appear only once in the hierarchy (for now, lol), and I handled the type remapping in the parent's slots using `unsafeChangeAttributeType` which doesn't change slot ordering.

Also, does the IR include specific slot numbers/indexes? I've only seen `getAttr` calls in IR that are [translated to slot numbers](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/interpreter.cpp#L801) at runtime [when `InterpreterState` is created](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/graph_executor.cpp#L390). So I don't think modifying the slot for an attribute after the IR has been generated will lead to problems. However, the return type of that `getAttr` still needs to be remapped after lowering.